src/schemas/__tests__/__snapshots__/Entity.test.js.snap

```diff
@@ -92,6 +92,100 @@ Object {
 }
 `;
 
+exports[`EntitySchema denormalization denormalizes recursive dependencies 1`] = `
+Object {
+  "draftedBy": Object {
+    "id": "456",
+    "reports": Array [
+      [Circular],
+    ],
+    "role": "manager",
+  },
+  "id": "123",
+  "publishedBy": Object {
+    "id": "456",
+    "reports": Array [
+      [Circular],
+    ],
+    "role": "manager",
+  },
+  "title": "Weekly report",
+}
+`;
+
+exports[`EntitySchema denormalization denormalizes recursive dependencies 2`] = `
+Object {
+  "draftedBy": Object {
+    "id": "456",
+    "reports": Array [
+      Object {
+        "draftedBy": "456",
+        "id": "123",
+        "publishedBy": "456",
+        "title": "Weekly report",
+      },
+    ],
+    "role": "manager",
+  },
+  "id": "123",
+  "publishedBy": Object {
+    "id": "456",
+    "reports": Array [
+      Object {
```

Comment:
When denormalizing entities, we're essentially doing a depth first search until we run out of nested schemas or we hit an entity we've already seen. To avoid infinite loop, we store the reference to the original (yet-to-be-denormalized) entity in the cache before recursing further.

For mutable objects, this results in us using the same exact reference which we're just mutating in `Entity.denormalize` and so creates a circular reference.
```js
report.draftedBy.reports[0] = originalReport
report === originalReport
```

For immutable objects, we point to the original reference, but during denormalization the entity's reference is changing.
```js
report.draftedBy.reports[0] = originalReport
report !== originalReport
```